planner/core/expression_rewriter.go

```diff
@@ -107,7 +107,11 @@ func (b *PlanBuilder) rewriteInsertOnDuplicateUpdate(ctx context.Context, exprNo
 // asScalar means whether this expression must be treated as a scalar expression.
 // And this function returns a result expression, a new plan that may have apply or semi-join.
 func (b *PlanBuilder) rewrite(ctx context.Context, exprNode ast.ExprNode, p LogicalPlan, aggMapper map[*ast.AggregateFuncExpr]int, asScalar bool) (expression.Expression, LogicalPlan, error) {
+	// capture the forUpdateRead mark
+	isForUpdateRead := b.isForUpdateRead
+	b.isForUpdateRead = false
 	expr, resultPlan, err := b.rewriteWithPreprocess(ctx, exprNode, p, aggMapper, nil, asScalar, nil)
+	b.isForUpdateRead = isForUpdateRead
```

Comment:
it seems without this change, TestIssue21498 still can work

and it seems subQuery in Projection will call `rewriteWithPreprocess` directly

https://github.com/pingcap/tidb/blob/f6cac2f831e54643d4367f6ab7a206f9ec7e1dcf/planner/core/logical_plan_builder.go#L1222

Reply:
Yeah, removed the capture now, at least not in this PR.